src/Umbraco.Web/PublishedCache/NuCache/DataSource/Lz4DictionaryOfPropertyDataSerializer.cs

```diff
@@ -0,0 +1,177 @@
+                    //  the 'current' value, and string.Empty should be used to represent the invariant or
+                    //  neutral values - PropertyData throws when getting nulls, so falling back to
+                    //  string.Empty here - what else?
+                    pdata.Culture = ReadStringObject(stream, true) ?? string.Empty;
+                    pdata.Segment = ReadStringObject(stream, true) ?? string.Empty;
+                    pdata.Value = ReadObject(stream);
+
+                    if ((map.Compress.Equals(NucachePropertyCompressionLevel.NucacheDatabase) || map.Compress.Equals(NucachePropertyCompressionLevel.SQLDatabase))
+                        && pdata.Value != null && pdata.Value is byte[] byteArrayValue)
+                    {
+                        //Compressed string
+                        switch (map.decompressionLevel)
+                        {
+                            case NucachePropertyDecompressionLevel.Lazy:
+                                pdata.Value = new LazyCompressedString(byteArrayValue);
+                                break;
+                            case NucachePropertyDecompressionLevel.NotCompressed:
+                                break;//Shouldn't be any not compressed
+                            case NucachePropertyDecompressionLevel.Immediate:
+                            default:
+                                pdata.Value = Encoding.UTF8.GetString(LZ4Pickler.Unpickle(byteArrayValue));
+                                break;
+                        }
+                    }
+                }
+
+                dict[key] = pdatas.ToArray();
+            }
+            return dict;
+        }
+
+        public void WriteTo(IDictionary<string, PropertyData[]> value, Stream stream)
+        {
+            // write properties count
+            PrimitiveSerializer.Int32.WriteTo(value.Count, stream);
+
+            // write each property
+            foreach (var (alias, values) in value)
+            {
+                var map = GetSerializationMap(alias);
+
+                // write alias
+                PrimitiveSerializer.String.WriteTo(map.MappedAlias ?? alias, stream);
+
+                // write values count
+                PrimitiveSerializer.Int32.WriteTo(values.Length, stream);
+
+                // write each value
+                foreach (var pdata in values)
+                {
+                    // everything that can be null is read/written as object
+                    //  even though - culture and segment should never be null here,
+                    //  see note in ReadFrom() method above
+                    WriteObject(pdata.Culture ?? string.Empty, stream);
+                    WriteObject(pdata.Segment ?? string.Empty, stream);
+
+                    //Only compress strings
+                    if (pdata.Value is string stringValue && pdata.Value != null && map.Compress.Equals(NucachePropertyCompressionLevel.NucacheDatabase)
```

Comment:
@nzdev I've merged this in and just cleaning up a few things to understand this all a little more. I've noticed that in this method we are only checking for this type NucachePropertyCompressionLevel.NucacheDatabase and not also NucachePropertyCompressionLevel.SQLDatabase but we are checking for both of those in the above ReadFrom method. Just wondering if this is on purpose or not?

Reply:
It's been a while, but the idea is potentially compressing twice. At the property level (NucachePropertyCompressionLevel.NucacheDatabase) as well as compressing the entire json (NucachePropertyCompressionLevel.SQLDatabase ). Compressing at the property level as well means the strings in bplustree are compressed.

Reply:
When reading back, we can decide if we want to decompress the properties or not before they go into bplustree.

Reply:
Ok cool, yep also wanted to check on what these meant as I'm adding some code comments:

* NucachePropertyCompressionLevel.NucacheDatabase = compression for properties at the nucache level (in memory + nucache.db)
* NucachePropertyCompressionLevel.SQLDatabase = compression for the entire document at the nucache DB level

I 'think' however, that on this line https://github.com/umbraco/Umbraco-CMS/pull/8400/files#diff-48a0f486bb419f4223394accbe88f205R85 we shouldn't be checking for NucachePropertyCompressionLevel.SQLDatabase? To me it seems like NucachePropertyCompressionLevel.SQLDatabase isn't really used quite yet since that is pretty much the only usage of it in the code. I could very well be missing something though :)

Reply:
Ah, sorry. It's where it's decompressed that makes a difference. NucachePropertyCompressionLevel has no effect on compressing the whole document.
NucachePropertyCompressionLevel.NucacheDatabase = compress the property before it's written to the database, store the compressed value in bplustree, lazy decompress the string on .ToString().

NucachePropertyCompressionLevel.SQLDatabase compress the property before it's written to the database, decompress when reading the document from the database, store the uncompressed value in bplustree. It's late in NZ, but I think you would only do this if you were not compressing the whole document.

Reply:
Ah, I think I understand a little more here, in the MsgPackContentNestedDataSerializer.Optimize we check for NucachePropertyCompressionLevel.SQLDatabase and we are compressing the property data if we aren't using the byte serialization. 

I'm just not sure there's any benefit in having that option since it seems storing the document as bytes with MessagePack is much better?

Reply:
NucachePropertyCompressionLevel.SQLDatabase is may be worth dropping. I think the thought at the time was it may be better to do all the decompression at boot.

Reply:
Cool just wanted to make sure i got it and makes sense now. Still cleaning up a few things, will post some revision changes when I'm done